CLI::IsMember

henryiii opened the pull request:
This implements `CLI::IsMember`, and deprecates the default `add_set*` options . This fixes the set lifetime and mutability issues in #170 and makes sets easy to modify after adding if you use a (shared) pointer. It also combines case insensitivity and underscore ignoring into validators, without the massive number of methods on App needed now. You can use any iterable that supports `std::begin`, `std::end`, and has `::value_type`, allowing a vector (for example) to be chosen to keep definition order - first item matches.

#### New features:

* Single new validator, no longer a massive list of functions in App
* Shared pointers supported, clear pointer vs. copy semantics
* Combinable custom filter functions
* All types support filter functions
* Other containers, including ordered containers allowed

#### Changes:

* The error for not being found in a set is now `CLI::ValidationError`
* All `add_set*` functions use the new backend, and are deprecated, planned removal of most of them in 1.9, and all of them after that. `mutable` versions may be removed even in 1.8.

#### Ideas:

A `->choices(...)` shortcut could be added, but the syntax is still up for debate, so not in this PR. Feel free to make suggestions.

#### Example of use:

```cpp
// Simple sets
app.add_option("--name", name)->check(CLI::IsMember({"tom", "harry"}));

// Sets can be modified later. `ignore_case` and `ignore_underscores` is addable.
std::set<std::string> names = {"tom", "harry"};
app.add_option("--name", name)->check(CLI::IsMember(&names, CLI::ignore_case));
names->insert("john");

// Shared pointers are supported, and different types. And ordered containers.
auto vals = std::make_shared<std::vector<int>>({1,2});
app.add_option("--val", val)->check(CLI::IsMember(&vals));
names->push_back(3);
```

You can use `std::shared_ptr` (that's internally what's used if you don't give a pointer-like). You can list `CLI::ignore_case` or `CLI::ignore_underscore` (or any other function that will be applied to both sides).

#### Still to do:

- [x] Deprecate old `add_set*` methods (can't be done on all of them because some are templates.)
- [x] Provide new versions of all the tests (coexist for a little while)
- Think about combined name + number style options (might not be part of this PR)
- [x] Add test with non-string function, like abs
- [x] Needs docs and examples
- [x] Fix `ignore_underscore` (no ending s) in the README.

henryiii committed:
Adding first draft of Sets

Use IsMember now

Using IsMember as backend for Set

Non-const validator backend

Move set tests to set

Clearer inits

phlptp commented on include/CLI/App.hpp:
```diff
@@ -1176,7 +1198,7 @@ class App {
     /// Get a pointer to subcommand by index
     App *get_subcommand(int index = 0) const {
-        if((index >= 0) && (index < subcommands_.size()))
+        if((index >= 0) && ((size_t)index < subcommands_.size()))
```
I suppose really you could use static cast and get rid of the (index>=0) check since a negative number cast into a large unsigned number.  I usually go the other way and cast the sizes to an int but It doesn't make that much difference.

henryiii replied:
In this case, since the index was already checked for `>= 0`, this should be valid. This should be valid for all sizes of ints. Though no one is likely to have something large enough to matter.

phlptp commented on include/CLI/Set.hpp:
```diff
@@ -0,0 +1,115 @@
+class Set {
+  protected:
+    using storage_type = std::vector<std::string>;
+    std::shared_ptr<storage_type> items_{std::make_shared<storage_type>()};
```
The question of how to enable mutable sets is going to be the tricky part here,  especially how to make it not unexpectedly mutable.

henryiii replied:
Sets are always mutable. Not sure why you would add a Set, then change it and not expect it to change. If you change it, it changes in the App. The thing this fixes is you can’t segfault by deleting the set.

phlptp replied:
I think my hesitancy is that when you see something passed by value it is usually copying(or moving) into it and it is then isolated from any changes that might be made at a higher level.    The shared_ptrs take care of lifetime issues with this, but it seems a little non-intuitive that any modifications to the set effect all previous uses of that set, even though it was passed by value.  

For example I have 1 use case where there are set of 5 different modes for one variable,  another has those 5 modes but adds an additional one.    Seeing sets are passed by value I might naively just add an element to the set and give it to the second option.  which would allow a hard to detect invalid set member in the first.  At the very least I am concerned that this will lead to some hard to detect and frustrating bugs on the part of users.  

I am actually not entirely sure what the best approach might be,  maybe take by pointer,  make a copy with the shared state,  then provide a deep copy function? .

henryiii replied:
Since CLI11 takes references as the main way of setting values, I felt that was pretty reasonable. However, since "mutability" is probably not as commonly needed, and should be clear to the user, I'm thinking that simplifying set and then taking two signatures, the set or a `shared_ptr` to a set, might be a good idea. The shared pointer version would clearly be mutable.

phlptp commented on include/CLI/Set.hpp:
```diff
@@ -0,0 +1,115 @@
+
+namespace CLI {
+
+class Set {
```
I wonder about the possibility of making set a class template, then you could have sets of `<int>`, or other things as long as they had a lexical cast or `>>` operator.  
```
Option *add_cli_set(std::string option_name,
                        T &member,   ///< The selected member of the set
                        Set<T> options, ///< The set of possibilities
                        std::string description = "") {
```

phlptp replied:
Along with a specialization for string that has the extra ignore_case and ignore_underscore functions

henryiii replied:
I thought about that, I just left it off for simplicity for now. I'll probably add it as part of making it work with enum outputs. However, since this is for "possible choices", I think requiring strings would not be too bad if it can't be done.

henryiii committed:
Drop shortcut

Tighten up classes a bit for MSVC

Check with GCC 4.8 too

henryiii committed:
Simpler templates, but more of them

Dropping more type safety for older compilers

Shortcut string set

henryiii committed:
Adding shortcut init

Making g++ 4.7 docker image happy

Fix Clang tidy issue with last commit

Adding one more shortcut, adding a couple of tests

henryiii committed:
Dropping dual pointer versions of code

henryiii committed:
Smarter shortcut syntax

henryiii committed:
Adding slighly faster choices

phlptp commented on include/CLI/Validators.hpp:
```diff
@@ -240,6 +264,116 @@ struct Range : public Validator {
+        tname = out.str();
+
+        func = [set, filter_fn](std::string &input) {
+            auto result = std::find_if(std::begin(set), std::end(set), [filter_fn, input](item_t v) {
```
This might look cleaner/simpler as a range based for loop instead of the find_if algorithm with a complex lambda, and you could collapse the else condition after result into the loop.  and a boolean found.    You also will only need to update the results if filter_fn actually contains something.

phlptp replied:
or just have the return directly from the loop if found

henryiii replied:
Done. I could split this into non-function versions too, so that `>>` would not be required on the type, but I think this is probably something that can be added later if anyone ever needs it.

henryiii committed:
Cleanup to make InMember simpler

henryiii committed:
Drop choices for now, adding some tests

phlptp commented on include/CLI/Validators.hpp:
```diff
@@ -240,6 +264,137 @@ struct Range : public Validator {
+
+    /// This checks to see if an item is in a set: pointer version. (Empty function)
+    template <typename T, enable_if_t<is_copyable_ptr<T>::value, detail::enabler> = detail::dummy>
+    explicit IsMember(T set)
```
may want to consider std::move(set) for the pointer and list syntax for the inner IsMember call, otherwise you will probably get a few copies of Set made unnecessarily.  Still never sure if the compilers optimize away some of the copies in contexts like this.

henryiii replied:
I think move is okay here.

phlptp commented on include/CLI/Validators.hpp:
```diff
@@ -240,6 +264,137 @@ struct Range : public Validator {
+                item_t a = v;
+                item_t b;
+                if(!detail::lexical_cast(input, b))
+                    throw ConversionError(input); // name is added later
```
Should this be a validationError rather than ConversionError, since it will occur in the validation context, and having one of those throw a conversion Error when someone is checking for a Validation Error could be confusing,  possibly making no difference whatsoever but if you can't convert to the appropriate type of a given set it is a conversionError that generates a validationError.  Which one should be set to the user?

henryiii replied:
The validator always runs first (that gives it the ability to affect the parse). So the following gives a conversion error:

```cpp
int myint;
app.add_option("--value", myint);
// Run with --value one
```

I didn't want the conversion error to change to a validation error here:

```cpp
int myint;
std::set<int> set = {1,2,3};
app.add_option("--value", myint)->check(CLI::IsMember(set));
// Run with --value one
```

But, both are true, and just saying "ValidationErrors" are stronger is probably fine.

You also have the odd mixed types that shouldn't happen but could:

```cpp
std::string mystr;
std::set<int> set = {1,2,3};
app.add_option("--value", mystr)->check(CLI::IsMember(set));
// Run with --value one
```
This fails only when checking the set and a ValidationError would fit better here, so I think I've convinced myself that ValidationError is better.

henryiii committed:
ValidationError is now always the error from a validator

phlptp commented on CMakeLists.txt:
```diff
@@ -36,7 +36,7 @@ if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
         add_definitions("/W4")
     else()
-        add_definitions(-Wall -Wextra -pedantic)
+        add_definitions(-Wall -Wextra -pedantic -Wpessimizing-move)
```
This new warning is not available on GCC, at least in MSYS2,  causes the compilation to fail, and isn't -Wpessimizing-move part of -Wall?    The failure is possibly due to some of gtest being C code and that isn't a valid warning in C.  

I had to remove it to get things to compile on MSYS2,  (not CLI11) itself just the test suite and everything.

henryiii replied:
Sorry, didn't mean to commit it! It's added by the other flags anyway, I think.

henryiii committed:
Support for other types of initializer lists, including enums

SkyToGround commented on include/CLI/App.hpp:
```diff
@@ -651,18 +651,8 @@ class App {
-        typeval += " in {" + detail::join(options) + "}";
-        opt->type_name(typeval);
+        Option *opt = add_option(option_name, member, std::move(description));
+        opt->check(IsMember{options});
```
👍

henryiii committed:
Factor out type utilities, single version of IsMember code

henryiii committed:
Adding a few tests for #224

henryiii committed:
Minor cleanup for Validation Error

henryiii committed:
Adding tests, moved deprecated tests

henryiii committed:
Docs updates
